<?xml version="1.0"?>

<!-- This is a ceiling fixed camera -->

<robot name="smocap" xmlns:xacro="http://www.ros.org/wiki/xacro">	

  <xacro:property name="PI" value="3.1415926535897931" />
  <xacro:property name="camera_x" value="1." />
  <xacro:property name="camera_y" value="1." />
  <xacro:property name="camera_z" value="2.9" />
  <xacro:property name="camera_r" value="0.01" />
  
  
  <gazebo>
    <static>true</static>
  </gazebo>
  
  <link name="camera_link">
    <!-- -->
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
	<box size="${camera_r} ${camera_r} ${camera_r}"/>
      </geometry>
    </collision>
    
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
	<box size="${camera_r} ${camera_r} ${camera_r}"/>
      </geometry>
      <material name="red"/>
      <cast_shadows>false</cast_shadows>
    </visual>
    
    <inertial>
      <mass value="1e-5" />
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6" />
      </inertial>
      <!-- -->
  </link>
    
  <link name="world" />
    
  <joint name="camera_joint"  type="fixed">
    <origin xyz="${camera_x} ${camera_y} ${camera_z}" rpy="0 ${PI/2} 0" />
    <parent link="world"/>
    <child link="camera_link"/>
  </joint>


  <link name="camera_optical_frame">
    <inertial>
      <mass value="0.01"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
    </inertial>
    <visual>
      <!-- bogus values for visual aesthetics -->
      <origin rpy="0 0 ${PI/2}" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.05" radius="0.02"/>
      </geometry>
      <material name="Blue"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.01 0.01 0.01"/>
      </geometry>
    </collision>
  </link>
  
  <joint name="camera_optical_frame_joint" type="fixed">
    <origin rpy="${-PI/2} 0.0 ${-PI/2}" xyz="0.0 0.0 0.0"/>
    <parent link="camera_link"/>
    <child link="camera_optical_frame"/>
  </joint>


  

  <!-- camera -->
  <xacro:property name="use_thermal_plugin" value="false" />

  <xacro:unless value="${use_thermal_plugin}">
  <gazebo reference="camera_link">
    <sensor type="camera" name="camera">
      <update_rate>30.0</update_rate>
      <camera name="head">
        <horizontal_fov>1.3962634</horizontal_fov>
        <image>
          <width>1600</width>
          <height>1200</height>
          <format>R8G8B8</format>
        </image>
        <clip>
          <near>0.02</near>
          <far>10</far>
        </clip>
        <noise>
          <type>gaussian</type>
          <!-- Noise is sampled independently per pixel on each frame.
               That pixel's noise value is added to each of its color
               channels, which at that point lie in the range [0,1]. -->
          <mean>0.0</mean>
          <stddev>0.007</stddev>
        </noise>
      </camera>
      <plugin name="camera_controller" filename="libgazebo_ros_camera.so">
        <alwaysOn>true</alwaysOn>
        <updateRate>0.0</updateRate>
        <cameraName>smocap/camera</cameraName>
        <imageTopicName>image_raw</imageTopicName>
        <cameraInfoTopicName>camera_info</cameraInfoTopicName>
        <frameName>camera_link</frameName>
        <hackBaseline>0.07</hackBaseline>
        <distortionK1>0.0</distortionK1>
        <distortionK2>0.0</distortionK2>
        <distortionK3>0.0</distortionK3>
        <distortionT1>0.0</distortionT1>
        <distortionT2>0.0</distortionT2>
      </plugin>
    </sensor>
  </gazebo>
  </xacro:unless>

  <xacro:if value="${use_thermal_plugin}">
    <gazebo reference="camera_optical_frame">
      <sensor type="camera" name="thermal_camera_sensor">
	<update_rate>30</update_rate>
	<camera name="head">
	  <horizontal_fov>1.3962634</horizontal_fov>
	  <image>
            <width>800</width>
            <height>800</height>
            <format>R8G8B8</format>
          </image>
          <clip>
            <near>0.02</near>
            <far>20</far>
          </clip>
	</camera>
	<plugin name="thermal_camera_controller" filename="libgazebo_ros_thermal_depth_camera.so">
          <alwaysOn>true</alwaysOn>
          <updateRate>0.</updateRate>
          <imageTopicName>/smocap/image_raw</imageTopicName>
          <cameraInfoTopicName>/smocap/camera_info</cameraInfoTopicName>
          <frameName>camera_link</frameName>
	</plugin>
      </sensor>
    </gazebo>
  </xacro:if>
  
</robot>
  
